### ISSUE TITLE
How to use bat as MANPAGER / PAGER? bat is disallowed as a pager.

### ISSUE THREAD

**Issue by salkin-mada** [sharkdp/bat#1433]
Thanks so much for making `bat`. It is a fantastic tool for making outputs human readable. 

I am not sure if this is just a misunderstanding on my end or if this is an issue. Therefore issue type is set to [question] for now.

When using bat as a MANPAGER with `export MANPAGER="sh -c 'col -bx | bat -l man -p'"`
`man cd` yields:
```
[bat error]: Use of bat as a pager is disallowed in order to avoid infinite recursion problems
man: command exited with status 1: sed -e '/^[[:space:]]*$/{ N; /^[[:space:]]*\n[[:space:]]*$/D; }' | LESS=-ix8RmPm Manual page cd(n) ?ltline %lt?L/%L.:byte %bB?s/%s..?e (END):?pB %pB\%.. (press h for help or q to quit)$PM Manual page cd(n) ?ltline %lt?L/%L.:byte %bB?s/%s..?e (END):?pB %pB\%.. (press h for help or q to quit)$ MAN_PN=cd(n) sh -c col -bx | bat -l man -p
```
This is maybe related to #652 ?
I also tried the above with `MANROFFOPT="-c"` no luck.
In addition I tried using `-bat -plman` as suggested in the issue. no luck.

I am on Arch Linux and installed bat via the community repo with `pacman -S bat`
```
❯ bat --version
bat 0.17.1
```
also I'm using `zsh`

**Comment by salkin-mada** [comment]
```
❯ man cd | bat
[bat error]: Use of bat as a pager is disallowed in order to avoid infinite recursion problems
```

**Comment by salkin-mada** [comment]
Aha this happens when env var `$PAGER` is set to `less`

**Comment by salkin-mada** [comment]
Okay my bad. I had exported PAGER to bat somewhere else on my system it seems. It now works with the line mentioned in the repo wiki. `MANPAGER="sh -c 'col -bx | bat -l man -p'"`

**Comment by cassepipe** [comment]
Any idea of why `MANPAGER="bat -l man -p " does not work well ?

**Comment by sghng** [comment]
> Any idea of why `MANPAGER="bat -l man -p "` does not work well ?

There are some control sequences used for decoration (e.g. bolding) in the output of `man`, seems like `col -bx` is needed to remove them, other wise, they'll be displayed as `^H` or something similar.

**Comment by CouscousPie** [comment]
@sghng From my experience, it is much more complex than that.
On my Debian simple `MANPAGER="batcat -l man -p` does work great with a bat from **apt**, while `MANPAGER="sh -c 'col -bx | batcat -l man -p'"` does not.

For some reason when I install man via brew on my Ubuntu I need
```export MANPAGER="sh -c 'sed -u -e \"s/\\x1B\[[0-9;]*m//g; s/.\\x08//g\" | bat -p -lman'"```
and the same command with batcat works for installation via apt. batcat from apt does however surprisingly not work with `MANPAGER="batcat -l man -p`.

All in all,  the variant with sed seems to be the most robust. Adding a distrobox in adds another layer of complexity and don't get my started with aliases like `bat="batcat"` ; they sometimes seem to work and sometimes not.

**Comment by HarrisonMc555** [comment]
My work computer is a MacBook and I run Ubuntu on my personal computer. I share the same Bash configuration between the two. I have observed the same as what has been mentioned here. To summarize:

* macOS
  * `MANPAGER="bat -l man -p" man ls` displays `^H` in various places
  * `MANPAGER="sh -c 'col -bx | bat -l man -p'" man ls'` displays a colored man page with no artifacts
* Ubuntu
  * `MANPAGER="bat -l man -p" man ls` displays a colored man page with no artifacts
  * `MANPAGER="sh -c 'col -bx | bat -l man -p'" man ls` displays some colors but also lots of unescaped color codes (e.g. `1mNAME0m`)

I have added a condition to my Bash configuration to use ~~`MANPAGER="bat -l man -p"`~~ `sh -c 'col -bx | bat -l man -p'"` on macOS and `MANPAGER="bat -l man -p" man ls` on Ubuntu.

**Comment by matrael** [comment]
@HarrisonMc555 
> I have added a condition to my Bash configuration to use `MANPAGER="bat -l man -p"` on macOS and `MANPAGER="bat -l man -p" man ls` on Ubuntu.

I'm curious, why would you set MANPAGER to the same command on both macOS and Ubuntu if it doesn't work on both? I can confirm that setting `MANPAGER="sh -c 'col -bx | bat -l man -p'"` works on macOS.

**Comment by HarrisonMc555** [comment]
@matrael ah that was a typo. I meant to say that I use _different_ values for `MANPAGER` based on the OS.
* macOS: `MANPAGER="sh -c 'col -bx | bat -l man -p'"`
* Ubuntu: `MANPAGER="bat -l man -p"`

**Comment by acook** [comment]
This thread has kinda turned into a support forum for mangled manpages and so it shows up when you search for the topic.

I ran into this same issue where sometimes manpages would have garbled text with improper syntax highlighting and bits of broken ANSI escape codes leftover. I tried a bunch of your suggestions and some others I found on SO.

After some experimentation I found `bat` can handle this itself using the `strip-ansi` option. As it implies, it will remove all ANSI escape sequences from its input. Then `bat` will apply its own effects.

```bash
# sh
export MANPAGER="bat --language=man --style=plain --strip-ansi=always"
```

```fish
# fish
set -gx MANPAGER "bat --language=man --style=plain --strip-ansi=always"
```

As for dealing with distros doing funny things with command names, this can be handled with a check in your shell's configuration. For example you can check for `batcat` and alias it to `bat` to make your configuration work across multiple systems:

```bash
# sh
command_exists() { command -v "$1" 1>&- 2>&-; }

if ! command_exists bat && command_exists batcat; then
  alias bat=batcat
fi
```

```fish
# fish
if ! type -q bat && type -q batcat
  function bat
    batcat $argv
  end
end
```

**Comment by CouscousPie** [comment]
`#Scroll to the bottom for final solution of Debian/Ubuntu systems`

I think there are a few separate issues mixed together here:

1. `bat` vs `batcat` is not just cosmetic in older/current Debian/Ubuntu packages as I mentioned in [my previous answer](https://github.com/sharkdp/bat/issues/1433#issuecomment-3298530339), that solved the problem with the best solution available at that time. Debian/Ubuntu apt packages have historically installed the executable as `batcat`, and `PAGER=batcat` does not necessarily hit the same recursion guard as `PAGER=bat`. In container tests, `PAGER=batcat batcat --paging=always ...` timed out/recursed instead of being treated like the guarded `bat` pager case.

2. `--strip-ansi=always` is a good newer solution, but not a universal one. It was added upstream on 2024-06-10 and first released in bat v0.25.0 on 2025-01-07. Debian first uploaded 0.25.0 to unstable on 2025-03-25; the first Debian stable release with it is Debian 13 trixie. Ubuntu first published it for Ubuntu 25.10 questing on 2025-04-22.

| Distro release | apt package version | Has `--strip-ansi`? |
|---|---:|---|
| Debian 11 bullseye | 0.12.1-6+b2 | no |
| Debian 12 bookworm | 0.22.1-4 | no |
| Debian 13 trixie | 0.25.0-2+b2 | yes |
| Debian testing/unstable | 0.26.1+dfsg-2+b1 | yes |
| Ubuntu 20.04 focal | 0.12.1-1ubuntu0.2 | no |
| Ubuntu 22.04 jammy | 0.19.0-1ubuntu0.1 | no |
| Ubuntu 24.04 noble | 0.24.0-1build1 | no |
| Ubuntu 25.10 questing | 0.25.0-2 | yes |
| Ubuntu 26.04 resolute | 0.25.0-5ubuntu1 | yes |


Evidence: Debian changelog has `0.25.0-1` on **2025-03-25** and `0.25.0-2` on **2025-03-30**; Debian package pages show bullseye/bookworm/trixie/forky/sid versions. Launchpad shows Ubuntu questing `0.25.0-2` on **2025-04-22**, resolute `0.25.0-5`, noble `0.24.0`, jammy `0.19.0`, focal `0.12.1`.

3. A shell alias like `alias bat=batcat` is not a reliable fix for `MANPAGER`, because `man` executes the pager command in its own shell context. A real executable shim/symlink is more reliable if you want a cross-distro `bat` command.

For new enough bat versions, this is probably the cleanest form:

```sh
export MANPAGER="bat --language=man --style=plain --strip-ansi=always"
```

On Debian/Ubuntu systems where only `batcat` exists:

```sh
export MANPAGER="batcat --language=man --style=plain --strip-ansi=always"
```

On older distro versions where `--strip-ansi` is not available, use a prefilter instead:

```sh
export MANPAGER="sh -c 'col -bx | batcat -l man -p'"
```

or, if ANSI escapes are the problem:

```sh
export MANPAGER="sh -c 'sed -u -e \"s/\\x1B\\[[0-9;]*m//g; s/.\\x08//g\" | batcat -p -lman'"
```

So I would not describe this as just “use `--strip-ansi` and alias `bat=batcat`”. That is correct only for sufficiently new `bat`, and aliases are shaky for this specific `MANPAGER` use case. Keep in mind these:

  - Avoid recommending alias bat=batcat for MANPAGER.
  - Avoid recommending --strip-ansi without saying it needs bat >= 0.25.0.
  - Prefer --paging=never inside a MANPAGER wrapper to avoid nested pager behavior.
  - Use batcat detection for Debian/Ubuntu.

All that being said, the best possible and stable solution is something like this:
```sh
#!/usr/bin/env sh

if command -v bat >/dev/null 2>&1; then
  bat_cmd=$(command -v bat)
elif command -v batcat >/dev/null 2>&1; then
  bat_cmd=$(command -v batcat)
else
  exec less
fi

if "$bat_cmd" --help 2>&1 | grep -q -- '--strip-ansi'; then
  exec "$bat_cmd" --language=man --style=plain --paging=never --strip-ansi=always
fi

col -bx | "$bat_cmd" --language=man --style=plain --paging=never
```
